source/javascripts/services/_stepSourceService.js.erb

```diff
@@ -1,143 +1,140 @@
+						gitURL: step.gitURL,
+						version: step.version
+					});
 
-				if (gitStep) {
-					step.defaultStepConfig = gitStep.defaultStepConfig;
-				}
+					if (gitStep) {
+						step.defaultStepConfig = gitStep.defaultStepConfig;
+					}
 
 				break;
-			default:
-				if (!source && !stepSourceService.defaultLibraryURL) {
-					throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_library] %>");
-				}
+				default:
+					if (!source && !stepSourceService.defaultLibraryURL) {
+						throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_library] %>");
+					}
 
-				if (id.length == 0) {
-					throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_id] %>");
-				}
+					if (id.length == 0) {
+						throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_id] %>");
+					}
 
-				step.libraryURL = source != null ? source : stepSourceService.defaultLibraryURL;
-				step.id = id;
-				step.version = version;
+					step.libraryURL = source != null ? source : stepSourceService.defaultLibraryURL;
+					step.id = id;
+					step.version = version;
 
-				var library = _.find(stepSourceService.libraries, {
-					url: step.libraryURL
-				});
+					var library = _.find(stepSourceService.libraries, {
+						url: step.libraryURL
+					});
 
-				if (!library) {
-					break;
-				}
-				var stepInLibrary = library.steps[step.id];
+					if (!library) {
+						break;
+					}
 
-				if (!stepInLibrary) {
-					throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_step_with_id_error_prefix] %>" + step.id);
-				}
+					if (!library.steps[step.id]) {
+						throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_step_with_id_error_prefix] %>" + step.id);
+					}
 
-				step.info = stepInLibrary.info;
+					var normalizedVersion = semverService.normalizeVersion(step.version);
+					var requestedVersion = semverService.resolveVersion(normalizedVersion, step.id, library);
 
-				var requestedVersion = step.version ? step.version : library.latestStepVersions[step.id];
+					if (!library.steps[step.id][requestedVersion]) {
+						throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_step_with_version_error_prefix] %>"+ requestedVersion);
+					}
 
-				if (!library.steps[step.id][requestedVersion]) {
-					throw new Error("<%= data[:strings][:step_source_service][:step_from_cvs][:no_step_with_version_error_prefix] %>"+ requestedVersion);
-				}
+					step.version = normalizedVersion;
+					step.defaultStepConfig = library.steps[step.id][requestedVersion].defaultStepConfig;
+				break;
+			}
 
-				step.version = requestedVersion;
-				step.defaultStepConfig = library.steps[step.id][requestedVersion].defaultStepConfig;
-		}
+			return step;
+		};
 
-		return step;
-	};
+		stepSourceService.versionsOfStep = function(step) {
+			if (!step.isLibraryStep()) {
+				return step.isLocal() ? null : [step.version];
```

Comment:
The code at some places counts on step version being undefined instead of null here:
* ~~stepSourceService: changeStepToVersion~~ nvm I see you update it in this PR
* workflows.slim: step list, version selector

Either use the orginal undefined in this case, or update the above mentioned code.

Reply:
I'd rather use `null` instead of `undefined`. Reason is: `undefined` is given by a runtime to indicate missing value, although `null` is explicit. Also has some better support for equality checks etc...

Reply:
@erosdome I've just tested the templates (we don't have e2e tests, so manually for now) and it seems `undefined` check is there for a good reason, since `null` means latest and that's ok, we should only hide sections when the version is `undefined`. I agree it's not super clear now, but let's say it's out of scope for this task.